### PR TITLE
[Model]Add DebertaV2ForSequenceClassification Model Support

### PR DESCRIPTION
**Pull request by yashaswipiplani**
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose
Added support for DebertaV2ForSequenceClassification models like: `meta-llama/Prompt-Guard-86M`, `mixedbread-ai/mxbai-rerank-xsmall-v1`, `microsoft/deberta-v2-xlarge-mnli`, etc

### CODE REVIEW

**Comment by alew3** [comment]
Would this be also compatible with meta-llama/Llama-Prompt-Guard-2-86M ? that would be very useful!

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM 5.1 on [Orb Cloud](https://orbcloud.dev))

## Summary

This PR adds support for DeBERTa v2 sequence classification models (`DebertaV2ForSequenceClassification`) to vLLM. It implements the full model architecture including DeBERTa's signature disentangled self-attention mechanism with content-to-position (c2p) and position-to-content (p2c) relative position encoding. The model is registered for cross-encoder/reranking tasks with `microsoft/deberta-v2-xlarge-mnli` as the example model.

## Architecture

The implementation adds a new model file `deberta_v2.py` alongside the existing `bert.py` pattern. The key architectural difference from BERT is DeBERTa's disentangled attention, which decomposes the self-attention computation into content-content, content-position, and position-content components using relative position embeddings.

**Registry integration:** Adds entries to `_CROSS_ENCODER_MODELS`, `tests/models/registry.py`, and `tests/models/test_registry.py`. Also updates the supported models documentation table.

## Issues

### 1. `ClassifierPooler` import will fail at runtime (critical)
**File:** `vllm/model_executor/models/deberta_v2.py`, line 89

```python
from vllm.model_executor.layers.pooler import Pooler, PoolingType, ClassifierPooler
```

`ClassifierPooler` does not exist in `vllm.model_executor.layers.pooler`. The available class is `ClassifierPoolerHead` in `vllm.model_executor.layers.pooler.seqwise.heads`, and the standard pattern (used by `BertForSequenceClassification`) is `DispatchPooler.for_seq_cls()` from `vllm.model_executor.layers.pooler.special`. This import will raise `ImportError` when loading any DeBERTa v2 model.

**Fix:** Follow the `bert.py` pattern:
```python
from vllm.model_executor.layers.pooler import DispatchPooler
# In __init__:
pooler_config = vllm_config.model_config.pooler_config
assert pooler_config is not None
self.pooler = DispatchPooler.for_seq_cls(
    pooler_config,
    pooling=self.deberta.pooler,
    classifier=self.classifier,
)
```

### 2. Merge conflict with main branch — `_CROSS_ENCODER_MODELS` renamed (critical)
**File:** `vllm/model_executor/models/registry.py`

The PR adds to `_CROSS_ENCODER_MODELS` which no longer exists on main — it was renamed to `_SEQUENCE_CLASSIFICATION_MODELS` (line 306 on current main). The PR's base appears to be from an older branch. This will cause a merge conflict and the registry entry won't work on current main.

### 3. Disentangled attention bypasses vLLM's attention infrastructure (warning)
**File:** `vllm/model_executor/models/deberta_v2.py`, `DebertaV2DisentangledSelfAttention.forward()`

The entire attention computation is implemented manually using raw `torch.matmul`, `torch.softmax`, etc. instead of vLLM's `EncoderOnlyAttention` or attention backend infrastructure. This means:
- No FlashAttention / flash-attn-v2 support → slower inference
- No paged attention / KV-cache optimizations
- No `@attn_type("encoder_only")` decorator for vLLM scheduling
- Manual mask handling instead of using vLLM's attention metadata

For a sequence classification model (short sequences, encoder-only), this is less critical than for a decoder model, but it's still a significant performance gap compared to the BERT implementation which uses `EncoderOnlyAttention`.

I understand DeBERTa's disentangled attention is fundamentally different from standard multi-head attention, so using vLLM's attention backends directly isn't straightforward. Consider at minimum:
- Adding `@attn_type("encoder_only")` to the model class
- Using `AttentionMetadata` properly for mask construction

### 4. Duplicate import (warning)
**File:** `vllm/model_executor/models/deberta_v2.py`

`PoolingMetadata` is imported twice:
- Line ~86: `from vllm.model_executor.pooling_metadata import PoolingMetadata`
- Line ~90: `from vllm.model_executor.pooling_metadata import PoolingMetadata`

### 5. Unused imports (warning)
**File:** `vllm/model_executor/models/deberta_v2.py`

Several imports are unused:
- `LinearBase` — not referenced anywhere in the file
- `DEFAULT_VOCAB_PADDING_SIZE` — not used
- `ParallelLMHead` — not used (the model uses `nn.Linear` for the classifier)
- `Attention`, `AttentionMetadata` — imported but attention is implemented manually
- `set_weight_attrs` — not used
- `F` — `torch.nn.functional` imported but not referenced
- `Union`, `List` — not used in type hints
- `QKVParallelLinear` — not used (Q/K/V projections use separate `ColumnParallelLinear`)

### 6. Manual attention batch handling is fragile (warning)
**File:** `vllm/model_executor/models/deberta_v2.py`, `disentangled_attention_bias()`

The per-sequence loop (`for b in range(batch_size)`) processes each batch element individually. This:
- Won't work correctly with vLLM's batching where sequences of different lengths are pre-padded and batched together
- The `_get_sequence_lengths()` method tries to get per-sequence lengths from `forward_context` which is a fragile pattern
- The single-sequence path shares the same `relative_pos` matrix across all batch elements

### 7. `build_relative_position` creates tensors on every forward pass (suggestion)
**File:** `vllm/model_executor/models/deberta_v2.py`, `DebertaV2Encoder.forward()`

`build_relative_position()` creates new relative position index tensors on every forward call. For a fixed max_position_embeddings, this should be computed once and cached (e.g., as a buffer).

### 8. Missing `is_pooling_model = True` and `SupportsCrossEncoding`/`SupportsQuant` interfaces (warning)
**File:** `vllm/model_executor/models/deberta_v2.py`

`BertForSequenceClassification` implements:
```python
class BertForSequenceClassification(nn.Module, SupportsCrossEncoding, SupportsQuant):
    is_pooling_model = True
```

But `DebertaV2ForSequenceClassification` is:
```python
class DebertaV2ForSequenceClassification(nn.Module):
```

Missing these interfaces may cause issues with vLLM's model dispatch and quantization support.

### 9. Weight loading is overly complex and fragile (warning)
**File:** `vllm/model_executor/models/deberta_v2.py`, `load_weights()`

The weight loading uses manual name mapping with string replacements and hardcoded skip lists. Compare with `BertForSequenceClassification` which uses the clean `AutoWeightsLoader` pattern:
```python
def load_weights(self, weights):
    loader = AutoWeightsLoader(self)
    return loader.load_weights(weights)
```

Consider using `WeightsMapper` for name mapping and `AutoWeightsLoader` for loading, following the pattern in `bert.py`.

### 10. Markdown table corruption in docs (suggestion)
**File:** `docs/models/supported_models.md`

The diff adds extra `|` characters and a `=======` merge conflict marker:
```
+                                                    |                       |
+=======
```
This will render incorrectly.

## Cross-file Impact

- **Registry (`registry.py`):** Needs rebase — `_CROSS_ENCODER_MODELS` was renamed to `_SEQUENCE_CLASSIFICATION_MODELS` on main.
- **Test registry (`tests/models/registry.py`):** Adds the example model for CI testing.
- **No changes to attention backends:** The model implements its own attention, so no impact on existing attention infrastructure.

## Assessment

**Request Changes** — The model cannot load or run in its current state due to the broken `ClassifierPooler` import. Additionally, the registry changes conflict with main branch. The architectural concerns (bypassing vLLM attention infrastructure, missing interfaces) should be addressed to ensure the model integrates properly with vLLM's serving layer and benefits from ongoing optimizations.

The disentangled attention implementation itself appears mathematically correct and faithful to the DeBERTa paper, which is the hardest part to get right. The integration issues are fixable by following the `bert.py` patterns more closely.

_This review was created by an AI assistant (OpenHands) on behalf of the Orb Cloud code review agent._

**Comment by JLiu4Coding** [comment]
Thanks for the work on this — DeBERTa support is definitely needed and this PR helped me a lot.

As this PR seems to havn't been active recently, I've opened a new issue (#42091) and an independent implementation as a draft PR (#42094) that addresses a few issues I found here (BaseModelOutput slicing crash, O(seq_len²) Python loops in disentangled attention, missing tensor parallelism). Happy to collaborate if you'd like to continue, otherwise feel free to close this in favour of the new PR.

**Comment by yashaswipiplani** [comment]
> Thanks for the work on this — DeBERTa support is definitely needed and this PR helped me a lot.
> 
> As this PR seems to havn't been active recently, I've opened a new issue (#42091) and an independent reimplementation as a draft PR (#42094) that addresses a few issues I found here (BaseModelOutput slicing crash, O(seq_len²) Python loops in disentangled attention, missing tensor parallelism). Happy to collaborate if you'd like to continue, otherwise feel free to close this in favour of the new PR.

Thanks for the new PR with limitations of this PR solved. I will close this one
